File: add-two-numbers.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x, nt=None):
        self.val = x
        self.next = nt


class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        head = l1
        flag = 0
        pre = ListNode(None)
        pre.next = l1
        while l1 or l2:
            if l1 and l2:
                val = l1.val + l2.val + flag
                l1.val = val % 10
            elif l1:
                val = l1.val + flag
                l1.val = val % 10
            elif l2:
                val = l2.val + flag
                l = ListNode(val % 10)
                pre.next = l
            flag = val / 10
            pre = pre and pre.next
            l1 = l1 and l1.next
            l2 = l2 and l2.next
        while flag:
            pre.next = ListNode(flag % 10)
            pre = pre.next
            flag /= 10
        return head


def test():
    l1 = ListNode(1, ListNode(2, ListNode(3)))
    l2 = ListNode(1, ListNode(2))
    head = Solution().addTwoNumbers(l1, l2)
    l = []
    while head:
        l.append(head.val)
        head = head.next
    print l
    assert l == [2, 4, 3]
    l1 = ListNode(12, ListNode(3123, ListNode(12)))
    l2 = ListNode(123, ListNode(21))
    head = Solution().addTwoNumbers(l1, l2)
    l = []
    while head:
        l.append(head.val)
        head = head.next
    print l
    assert l == [5, 7, 7, 2, 3]

if __name__ == '__main__':
    test()

File: algorithms/medium/add-two-numbers.c
#include <stdio.h>
#include <stdlib.h>

struct ListNode {
    int val;
    struct ListNode *next;
};

struct ListNode* newListNode(int val, struct ListNode* next) {
    struct ListNode* node = malloc(sizeof node);
    node->val = val;
    node->next = next;
    return node;
}
 
struct ListNode* addTwoNumbers(struct ListNode* l1, struct ListNode* l2) {
    if(l1 == NULL) {
        return l2;
    }
    if(l2 == NULL) {
        return l1;
    }
    struct ListNode* l3 = NULL;
    struct ListNode* head = NULL;
    int f = 0;
    int val = 0;
    while(l1 != NULL && l2 != NULL) {
        val = l1->val + l2->val + f;
        f = val / 10;
        if(head == NULL) {
            head = malloc(sizeof *head);
            head->val = val % 10;
            head->next = NULL;
            l3 = head;
        } else {
            struct ListNode* node = malloc(sizeof (struct ListNode));
            node->val = val % 10;
            node->next = NULL;
            head->next = node;
            head = head->next; 
        }
        l1 = l1->next;
        l2 = l2->next;
    }
    while(l1 != NULL) {
        val = l1->val + f;
        f = val / 10;
        struct ListNode* node = malloc(sizeof (struct ListNode));
        node->val = val % 10;
        node->next = NULL;
        head->next = node;
        head = head->next;
        l1 = l1->next;
    }
    while(l2 != NULL) {
        val = l2->val + f;
        f = val / 10;
        struct ListNode* node = malloc(sizeof (struct ListNode));
        node->val = val % 10;
        node->next = NULL;
        head->next = node;
        head = head->next;
        l2 = l2->next;
    }
    if(f != 0) {
        struct ListNode* node = malloc(sizeof (struct ListNode));
        node->val = f;
        node->next = NULL;
        head->next = node;
        head = head->next;
    }
    return l3;
}

int main() {
    struct ListNode* l1 = newListNode(2, newListNode(4, newListNode(3, NULL)));
    struct ListNode* l2 = newListNode(5, newListNode(6, newListNode(8, newListNode(9, NULL))));
    struct ListNode* l3 = addTwoNumbers(l1, l2);
    while(l3 != NULL) {
        printf("%d ", l3->val);
        l3 = l3->next;
    }
}
